pkg/app/server.go

```diff
@@ -323,10 +336,104 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
 	if err := g.Run(); err != nil {
 		return fmt.Errorf("run server group error: %v", err)
 	}
+
 	klog.InfoS("Exited")
 	return nil
 }
 
+func pollForCacheUpdates(
+	ctx context.Context,
+	opts *options.Options,
+	discovererInstance *discovery.CRDiscoverer,
+	storeBuilder *store.Builder,
+	m *metricshandler.MetricsHandler,
+) {
+	// The interval at which we will check the cache for updates.
+	t := time.NewTicker(discovery.Interval)
+	// Track previous context to allow refreshing cache.
+	olderContext, olderCancel := context.WithCancel(ctx)
+	// Prevent context leak (kill the last metric handler instance).
+	defer olderCancel()
+	generateMetrics := func() {
+		// Generate config for when none of the CRDs were present during callers' initialization.
+		// This is to ensure that the config does not throw an EOF error due to coming up empty.
+		customConfig, err := resolveCustomResourceConfig(opts)
```

Comment:
this is a very expensive operation we shouldn't do it every tick. Rather the config should only be updated here when it is updated by the users. Manuel recently added some hot reloading for the custom-state config that we could maybe hook here.

Reply:
Assuming you are referring to [this](https://github.com/kubernetes/kube-state-metrics/commit/06268ab5ef9143038451117718c3bc90d9032492) change, any config modification will effectively restart KSM, and this with it. I've taken this out of the function since we need to get this only once (because any change will trigger a KSM restart).